vck-openid-ktor/src/commonMain/kotlin/at/asitplus/wallet/lib/ktor/openid/Extensions.kt

```diff
@@ -1,12 +1,60 @@
 package at.asitplus.wallet.lib.ktor.openid
 
+import at.asitplus.openid.OpenIdConstants.Errors.USE_DPOP_NONCE
+import at.asitplus.wallet.lib.oidvci.OAuth2Error
+import io.ktor.client.call.*
+import io.ktor.client.statement.*
+import io.ktor.http.*
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.CoroutineStart
 import kotlinx.coroutines.Deferred
 import kotlinx.coroutines.async
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.InvocationKind
+import kotlin.contracts.contract
 
 fun <T> CoroutineScope.lazyDeferred(
     block: suspend CoroutineScope.() -> T,
 ): Lazy<Deferred<T>> = lazy {
     async(start = CoroutineStart.LAZY) { block() }
 }
+
+/** Intermediate class to perform error handling on ktor responses, see [onFailure] and [onSuccess]. */
+sealed class IntermediateResult<R> {
+    class Success<R>(val httpResponse: HttpResponse) : IntermediateResult<R>()
+    class Failure<R>(val result: R) : IntermediateResult<R>()
+}
+
+/** Helper method to perform error handling on ktor responses, see [onSuccess]. */
+@OptIn(ExperimentalContracts::class)
+suspend inline fun <reified R> HttpResponse.onFailure(
```

Comment:
do we maybe want to add a parameter (with a default catch-all value) to restrict the error rage or the specific error we want to handle? as in:
`onFailure(400) {…}` or `onFailure(400..500/*IntRange*/) {…}`. Of course, we'd then need a third "branch" to handle all other errors, so what is now `onFailure` as an extension on HttpResponse would also need to be an extension on som other intermediate object in addition to handling all errors from HttpResponse directly. Same goes for `onSuccess`